qubes-gui-client/src/buffer.rs

```diff
@@ -268,25 +279,79 @@ impl<T: VchanMock + 'static> RawMessageStream<T> {
+                                Err(e) => {
+                                    self.state = ReadState::Error;
+                                    break Err(e.into());
+                                }
+                            }
+                        }
+                        Kind::Daemon { .. } => self.state = ReadState::Negotiating,
+                    },
                     vchan::Status::Disconnected => {
                         break self.fail(ErrorKind::Other, "vchan connection refused");
                     }
                 },
                 ReadState::Error => break self.fail(ErrorKind::Other, "Already in error state"),
-                ReadState::ReadingXConf if ready >= SIZE_OF_XCONF => {
-                    self.xconf = self.recv_struct()?;
-                    self.state = ReadState::ReadingHeader;
-                }
+                ReadState::Negotiating => match self.kind {
+                    Kind::Agent if ready >= SIZE_OF_XCONF => {
+                        let xconf: qubes_gui::XConfVersion = self.recv_struct()?;
+                        let (daemon_major, daemon_minor) =
+                            (xconf.version >> 16, xconf.version & 0xFFFF);
+                        if qubes_gui::PROTOCOL_VERSION_MAJOR == daemon_major
+                            && qubes_gui::PROTOCOL_VERSION_MINOR >= daemon_minor
+                            && daemon_minor >= 4
+                        {
+                            self.xconf = xconf;
+                            self.state = ReadState::ReadingHeader
+                        } else {
+                            self.state = ReadState::Error;
+                            break Err(Error::new(ErrorKind::InvalidData,
+                                            format!(
+                                                "Version negotiation failed: their version is {}.{} but ours is {}.{}",
+                                                daemon_major, daemon_minor,
+                                                qubes_gui::PROTOCOL_VERSION_MAJOR,
+                                                qubes_gui::PROTOCOL_VERSION_MINOR,
+                                                )));
+                        }
+                    }
+                    Kind::Daemon if ready >= 4 => {
+                        let version: u32 = self.recv_struct()?;
+                        let (major, minor) = (version >> 16, version & 0xFFFF);
+                        if major == qubes_gui::PROTOCOL_VERSION_MAJOR {
+                            let version = version.min(qubes_gui::PROTOCOL_VERSION_MINOR);
+                            self.xconf.version = version;
+                            if version >= 4 {
+                                self.vchan.send(self.xconf.as_bytes())?
+                            } else {
+                                self.vchan.send(self.xconf.xconf.as_bytes())?
+                            }
```

Comment:
If this fails, the state machine won't enter the `Error` state. This is not the first time there's a bug like this, so I'd suggest structuring the code to make it impossible to write more like it. 

For example, you could move all the logic to a `read_message_internal` function, and then have `read_message` itself do
```
pub fn read_message<'a>(&'a mut self) -> io::Result<Option<ClaimableBuffer<'a, T>>> {
    match self.read_message_internal() {
        Ok(v) => Ok(v),
        Err(e) => {
            self.state = ReadState::Error;
            Err(e)
        }
    }
}
```

Reply:
If I remember correctly from when I tried this myself, you indeed do need that if you write the `internal` as a closure, but it isn't necessary if you write it as an actual function.

Reply:
I get

```
    Checking qubes-gui-client v0.1.0 (/home/user/qubes-builder/qubes-src/gui-rust/qubes-gui-client)
error[E0506]: cannot assign to `self.state` because it is borrowed
   --> qubes-gui-client/src/buffer.rs:360:17
    |
356 |     pub fn read_message<'a>(&'a mut self) -> io::Result<Option<Buffer<'a>>> {
    |                         -- lifetime `'a` defined here
357 |         match self.read_message_internal() {
    |               ---------------------------- borrow of `self.state` occurs here
358 |             Ok(v) => Ok(v),
    |                      ----- returning this value requires that `*self` is borrowed for `'a`
359 |             Err(e) => {
360 |                 self.state = ReadState::Error;
    |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ assignment to borrowed `self.state` occurs here

For more information about this error, try `rustc --explain E0506`.
error: could not compile `qubes-gui-client` due to previous error
```

Can you post some code that works?  Alternatively, should this be rewritten in a push style with callbacks?

Reply:
Merged (with minor cleanups on commit; you got the author line since it was your idea and did almost all the work).